src/test/scala/org/kframework/minikore/NodeTests.scala

```diff
@@ -0,0 +1,205 @@
+package org.kframework.minikore
+
+import org.junit.Test
+import org.kframework.minikore.interfaces.pattern._
+import org.kframework.minikore.interfaces.tree._
+import org.kframework.minikore.implementation.DefaultBuilders
+import org.kframework.minikore.interfaces.build
+
+
+class NodeTest {
+
+  object TestFunctions {
+
+    def size(p: Pattern): Int = p match {
+      case Node(c: Seq[Pattern]) => c.map(size).sum
+      case Leaf(_) => 1
+    }
+
+    def getNode0Count(p: Pattern): Int = p match {
+      case Node0() => 1
+      case Node(c: Seq[Pattern]) => c.map(getNode0Count) sum
```

Comment:
BTW, I'm not trying to be ironic here, I literally want to understand this.  So if you know any any good reading on this issue, please let us know @cos .